packages/snaps-utils/src/structs.ts

```diff
@@ -289,3 +300,68 @@ export function getStructErrorMessage<Type, Schema>(
 
   return formattedFailures.join('\n');
 }
+
+/**
+ * Validate a union struct, and throw readable errors if the value does not
+ * satisfy the struct. This is useful for improving the error messages returned
+ * by `superstruct`.
+ *
+ * @param value - The value to validate.
+ * @param struct - The `superstruct` union struct to validate the value against.
+ * This struct must be a union of object structs, and must have at least one
+ * shared key to validate against.
+ * @param structKey - The key to validate against. This key must be present in
+ * all the object structs in the union struct, and is expected to be a literal
+ * value.
+ * @throws If the value does not satisfy the struct.
+ * @example
+ * const struct = union([
+ *   object({ type: literal('a'), value: string() }),
+ *   object({ type: literal('b'), value: number() }),
+ *   object({ type: literal('c'), value: boolean() }),
+ *   // ...
+ * ]);
+ *
+ * // At path: type — Expected the value to be one of: "a", "b", "c", but received: "d".
+ * validateUnion({ type: 'd', value: 'foo' }, struct, 'type');
+ *
+ * // At path: value — Expected a value of type string, but received: 42.
+ * validateUnion({ type: 'a', value: 42 }, struct, 'value');
+ */
+export function validateUnion<Type, Schema extends readonly Struct<any, any>[]>(
+  value: unknown,
+  struct: Struct<Type, Schema>,
+  structKey: keyof Type,
+) {
+  assert(struct.schema.length > 0, 'Expected a non-empty array of structs.');
+
+  const keyUnion = struct.schema.map(
+    (innerStruct) => innerStruct.schema[structKey],
+    // This is guaranteed to be a non-empty array by the assertion above. We
+    // need to cast it since `superstruct` requires a non-empty array of structs
+    // for the `union` struct.
+  ) as NonEmptyArray<Struct>;
+
+  const key = superstructType({
+    [structKey]: union(keyUnion),
+  });
+
+  const [keyError] = validate(value, key);
+  if (keyError) {
+    throw new Error(getStructFailureMessage(key, keyError.failures()[0]));
+  }
+
+  // At this point it's guaranteed that the value is an object, so we can safely
+  // cast it to a Record.
+  const objectValue = value as Record<PropertyKey, unknown>;
+  const objectStruct = struct.schema.find((innerStruct) =>
+    is(objectValue[structKey], innerStruct.schema[structKey]),
```

Comment:
For my understanding, how does `is` pass, but then `validate` may fail later?

Reply:
`structKey` is the common property in each object (`type` in the example), and this only checks if the "inner" struct's common property matches the value's common property (i.e., `value.type === struct.type`). Other properties of the value may still be invalid.

Reply:
Ah, makes sense!